src/library/features/crates/cratehierarchy.cpp

```diff
@@ -0,0 +1,529 @@
+
+    // get the last item (childId)
+    childId = CrateId(ids.back().toInt());
+    if (ids.size() > 1) {
+        // get the second to last item (parentId)
+        parentId = CrateId(ids.at(ids.size() - 2));
+    } else {
+        // if there isn't one return false
+        return false;
+    }
+    return true;
+}
+
+
+bool CrateHierarchy::initClosureForCrate(CrateId id) const {
+    FwdSqlQuery query(
+      m_database, QString(
+        "INSERT INTO %1 "
+        "VALUES(:parent, :child, 0)").arg(
+          CRATE_CLOSURE_TABLE));
+
+    if (!query.isPrepared()) {
+        return false;
+    }
+
+    // closure dependacy
+    query.bindValue(":parent", id);
+    query.bindValue(":child", id);
+    if (!query.execPrepared()) {
+        return false;
+    }
+
+    return true;
+}
+
+bool CrateHierarchy::insertIntoClosure(CrateId parent, CrateId child) const {
+    FwdSqlQuery query(
+      m_database, QString(
+        "INSERT INTO %1(%2, %3, %4) "
+        "SELECT p.%2, c.%3, p.%4 + c.%4 + 1 "
+        "FROM %1 p, %1 c "
+        "WHERE p.%3 = :parent AND c.%2 = :child").arg(
+          CRATE_CLOSURE_TABLE,
+          CLOSURE_PARENTID,
+          CLOSURE_CHILDID,
+          CLOSURE_DEPTH));
+
+    if (!query.isPrepared()) {
+        return false;
+    }
+
+    query.bindValue(":parent", parent);
+    query.bindValue(":child", child);
+
+    if (!query.execPrepared()) {
+        return false;
+    }
+
+    return true;
+}
+
+void CrateHierarchy::deleteCrate(CrateId id) const {
+    {
+        FwdSqlQuery query(
+          m_database, QString(
+            "DELETE FROM cratePath WHERE crateId = :id"));
+        query.bindValue(":id", id);
+        if (!query.isPrepared()) {
+            return;
+        }
+        if (!query.execPrepared()) {
+            return;
+        }
+    }
+
+    {
+        FwdSqlQuery query(
+          m_database, QString(
+            "DELETE FROM crateClosure WHERE childId = :id"));
+        query.bindValue(":id", id);
+        if (!query.isPrepared()) {
+            return;
+        }
+        if (!query.execPrepared()) {
+            return;
+        }
+    }
+}
+
+QString CrateHierarchy::formatQueryForTrackIdsByCratePathLike(const QString& cratePathLike) const {
+    FieldEscaper escaper(m_database);
+    QString escapedArgument = escaper.escapeString(kSqlLikeMatchAll + cratePathLike + kSqlLikeMatchAll);
+
+    return QString(
+        "SELECT DISTINCT %1 FROM %2 "
+        "JOIN %3 ON %4=%5 "
+        "WHERE %6 LIKE %7 "
+        "ORDER BY %1").arg(
+            CRATETRACKSTABLE_TRACKID,
+            CRATE_TRACKS_TABLE,
+            CRATE_PATH_TABLE,
+            PATHTABLE_CRATEID,
+            CRATETRACKSTABLE_CRATEID,
+            PATHTABLE_NAME_PATH,
+            escapedArgument);
+}
+
+bool CrateHierarchy::canBeRenamed(const QString& newName,
+                                  const Crate& crate,
+                                  const CrateId parentId) const {
+    if (parentId.isValid()) {
```

Comment:
@Be-ing I'd like for the code to forbid it so the filter would work better, since it's based on name and putting two crates with the same name in the same path seems to me it could yield really confusing results.